cub/test/catch2_test_device_radix_sort_keys.cu

```diff
@@ -0,0 +1,471 @@
+ *
+ ******************************************************************************/
+
+#include <thrust/device_vector.h>
+#include <thrust/functional.h>
+#include <thrust/host_vector.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/memory.h>
+#include <thrust/scatter.h>
+#include <thrust/transform.h>
+
+#include <cub/device/device_radix_sort.cuh>
+#include <cub/util_type.cuh>
+
+#include <cuda/std/type_traits>
+
+#include <algorithm>
+#include <limits>
+
+#include "catch2_sort_helper.cuh"
+#include "catch2_test_helper.h"
+#include "catch2_test_launch_helper.h"
+
+// %PARAM% TEST_LAUNCH lid 0:1:2
+
+DECLARE_LAUNCH_WRAPPER(cub::DeviceRadixSort::SortKeys, sort_keys);
+DECLARE_LAUNCH_WRAPPER(cub::DeviceRadixSort::SortKeysDescending, sort_keys_descending);
+
+// %PARAM% TEST_KEY_BYTES key_bytes 8:16:32:64
+
+// TODO:
+// - int128
+// - uint128
+
+// The unsigned integer for the given byte count should be first:
+#if TEST_KEY_BYTES == 8
+using key_types = c2h::type_list<cuda::std::uint8_t, cuda::std::int8_t, bool, char>;
+#elif TEST_KEY_BYTES == 16
+// clang-format off
+using key_types = c2h::type_list<
+    cuda::std::uint16_t
+  , cuda::std::int16_t
+#ifdef TEST_HALF_T
+  , half_t
+#endif
+#ifdef TEST_BF_T
+  , bfloat16_t
+#endif
+  >;
+// clang-format on
+#elif TEST_KEY_BYTES == 32
+using key_types = c2h::type_list<cuda::std::uint32_t, cuda::std::int32_t, float>;
+#elif TEST_KEY_BYTES == 64
+using key_types = c2h::type_list<cuda::std::uint64_t, cuda::std::int64_t, double>;
+#endif
+
+// Used for testing begin/end bit slicing
+template <typename T>
+struct supports_bit_window
+{
+  static constexpr bool value = cuda::std::is_unsigned<T>::value && !cuda::std::is_same<bool, T>::value;
+};
+using bit_window_key_types = c2h::copy_if<key_types, c2h::trait<supports_bit_window>>;
+
+// Used for testing floating-point specific behavior
+using fp_key_types = c2h::copy_if<key_types, c2h::trait<cuda::std::is_floating_point>>;
+
+// Used for tests that just need a single type for testing:
+using single_key_type = c2h::type_list<c2h::get<0, key_types>>;
+
+// Index types used for NumItemsT testing
+using num_items_types =
+  c2h::type_list<cuda::std::uint32_t, cuda::std::int32_t, cuda::std::uint64_t, cuda::std::int64_t>;
+
+// Helpers to assist with specifying default args to DeviceRadixSort API:
+template <typename T>
+constexpr int begin_bit()
+{
+  return 0;
+}
+
+template <typename T>
+constexpr int end_bit()
+{
+  return static_cast<int>(sizeof(T) * CHAR_BIT);
+}
+
+CUB_TEST("DeviceRadixSort::SortKeys: basic testing", "[keys][radix][sort][device]", key_types)
+{
+  using key_t = c2h::get<0, TestType>;
+
+  // Test different magnitudes:
+  const int power_of_two = GENERATE(0, 5, 10, 20);
+  // Test slightly less than a power of two, exactly a power of two, and slightly more:
+  const int offset            = GENERATE(-1, 0, 3);
+  const std::size_t num_items = (1 << power_of_two) + offset;
```

Comment:
suggestion: I'd recommend the approach from other tests, where `num_items` is randomly generated. When we introduce fuzz testing, this will lead to better coverage than power-of-two neighborhood.